### PR TITLE
Converted LpNorm, GcNorm and Whitening transforms into transformers/estimators…

### LINKED ISSUE #959: Convert LpNorm, GcNorm and Whitening Transforms to transformers/estimators.
In this work item, following transforms will be converted into transformer/estimators.

- LpNormNormalizerTransform
- GcnTransform
- WhiteningTransform

### PR DESCRIPTION
**Pull request by zeahmed**
This PR fixes #959

### CODE REVIEW

File: src/Microsoft.ML.Transforms/WrappedGcnTransformers.cs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                return new GcNormalizer(env, pairs.ToArray(), _subMean, _useStdDev, _scale);
+            }
+        }
+
+        /// <include file='doc.xml' path='doc/members/member[@name="LpNormalize"]/*'/>
+        /// <param name="input">The column to apply to.</param>
+        /// <param name="subMean">Subtract mean from each value before normalizing.</param>
+        /// <param name="useStdDev">Normalize by standard deviation rather than L2 norm.</param>
+        /// <param name="scale">Scale features by this value.</param>
+        public static Vector<float> GcNormalize(this Vector<float> input,
```
Comment (Zruty0) [inline src/Microsoft.ML.Transforms/WrappedGcnTransformers.cs]: >GcNormalize [](start = 36, length = 11)

While 'Lp' is fine (it's a mathematical notation L_p), 'Gc' is probably not. 'GlobalContrast' ? Does it even make sense to have both as extensions though? They seem to share a lot in common

File: src/Microsoft.ML.Transforms/WrappedWhiteningTransformer.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        /// <include file='doc.xml' path='doc/members/member[@name="Whitening"]/*'/>
+        /// <param name="input">The column to apply to.</param>
+        /// <param name="kind">Whitening kind (PCA/ZCA).</param>
+        /// <param name="eps">Scaling regularizer.</param>
+        /// <param name="maxRows">Max number of rows.</param>
+        /// <param name="saveInverse">Whether to save inverse (recovery) matrix.</param>
+        /// <param name="pcaNum">PCA components to retain.</param>
+        public static Vector<float> Whiten(this Vector<float> input,
```
Comment (Zruty0) [inline src/Microsoft.ML.Transforms/WrappedWhiteningTransformer.cs]: >Whiten [](start = 36, length = 6)

this sounds like a wacky thing to do to an arbitrary float array, is it not?
Do you think there is a good name that's more elaborate?

And if not, maybe consider not having the pigstension at all?

File: test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs [hunk 3]
```diff
@@ -431,5 +432,39 @@ public void Tokenize()
...
+
+        [Fact]
+        public void LpGcNormAndWhitening()
+        {
+            var env = new ConsoleEnvironment(seed: 0);
+            var dataPath = GetDataPath("generated_regression_dataset.csv");
+            var dataSource = new MultiFileSource(dataPath);
+
+            var reader = TextLoader.CreateReader(env,
+                c => (label: c.LoadFloat(11), features: c.LoadFloat(0, 10)),
```
Comment (Zruty0) [inline test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs]: >11 [](start = 41, length = 2)

these are 0-based, are you sure?

File: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs [hunk 4]
```diff
@@ -88,5 +88,40 @@ public void TextTokenizationWorkout()
             CheckEquality("Text", "tokenized.tsv");
             Done();
         }
+
+        [Fact]
+        public void LpGcNormAndWhiteningWorkout()
```
[the remark this answers is not in the document]
Comment (Zruty0) [reply test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs]: also move out of text featurizer tests

---
In reply to: [219011012](https://github.com/dotnet/machinelearning/pull/961#discussion_r219011012) [](ancestors = 219011012)

File: test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs (hunk 3 above)
Comment (zeahmed) [reply test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs]: These is i think.

features are from 0-10 and label is 11. Do you mean something else?

---
In reply to: [219263470](https://github.com/dotnet/machinelearning/pull/961#discussion_r219263470) [](ancestors = 219263470)

Comment (Zruty0) [reply test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs]: Yea, it's my mistake

---
In reply to: [219289826](https://github.com/dotnet/machinelearning/pull/961#discussion_r219289826) [](ancestors = 219289826,219263470)

File: src/Microsoft.ML.Transforms/WrappedGcnTransformers.cs (hunk 1 above)
Comment (zeahmed) [reply src/Microsoft.ML.Transforms/WrappedGcnTransformers.cs]: I think they have different set of parameters and mixing them into one would not look good. I have renamed `GcNormalize` to `GlobalContrastNormalize`.

---
In reply to: [219260214](https://github.com/dotnet/machinelearning/pull/961#discussion_r219260214) [](ancestors = 219260214)

**Pull request merged**